HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,221 @@
+potential_parents = rdd.filter(lambda x: x[1] == filter_distance).partitionBy(256) #667
+
+potential_parents_with_children = potential_parents.join(full_rdd2).mapValues(lambda x: x[1]) #667, p
+
+parents_of_child = potential_parents_with_children.filter(lambda x: child in x[1]).keys() #364, pp
+
+#Write out middle nodes
+middle_step = parents_of_child.collect()
+
+single_parent = page_names.filter(lambda (K, V): V == int(middle_step[0])).collect()
+
+ 
+
+all_steps = ['Kevin_Bacon', single_parent[0][0], 'Harvard_University']
+
+print "KB Shortest Path:", all_steps
+
+pickle.dump(single_parent, open( "KB_HARVARD.pkl", "wb" ) )
+
+
+
+
+############################################# Harvard to Kevin #############################
+
+start_node = Harvard_University
+end_node = Kevin_Bacon
+
+start = [(start_node, 0)] 
+rdd = sc.parallelize(start).partitionBy(256)
+i = 0
+
+num_runs = 4
+
+while i < num_runs:
+    assert copartitioned(full_rdd2, rdd)
+    
+    rdd_parent = rdd.filter(lambda x: x[1] == i).join(full_rdd2).mapValues(lambda x: x[1])
+    assert copartitioned(rdd, rdd_parent)
+    
+    rdd_children = rdd_parent.flatMap(lambda x: x[1]).map(lambda x: (x, 1)).subtractByKey(rdd).distinct().partitionBy(256)
+    print 'distinct/subtract', i
+    assert copartitioned(rdd, rdd_children)
+    
+    rdd_intermediate = rdd_children.map(fun(i), True)
+    print 'change radii', i
+    assert copartitioned(rdd, rdd_children)
+    
+    rdd = rdd.union(rdd_intermediate).cache()
+    print 'union:', i
+    assert copartitioned(rdd, rdd_intermediate)
+
+    i += 1 #iterator may not work on cluster
+    print 'end of iter', i
+
+#Write Out Information
+
+#Write Out Node Information for complete graph
+#node_distances = rdd.collect()
+#pickle.dump(node_distances, open( "Harvard_Distances_KB.pkl", "wb" ) )
+
+
+#Establish know initial parameters
+child_distance = 3
+filter_distance = child_distance - 1
+child = Kevin_Bacon
+
+
+potential_parents = rdd.filter(lambda x: x[1] == filter_distance).partitionBy(256) 
+
+potential_parents_with_children = potential_parents.join(full_rdd2).mapValues(lambda x: x[1]) 
+
+parents_of_child = potential_parents_with_children.filter(lambda x: child in x[1]).keys() 
+
+#Write out middle nodes
+middle_step = parents_of_child.collect()
+
+one_behind_harvard = page_names.filter(lambda (K, V): V == int(middle_step[1])).collect()
+
+#print 'harvard middle', middle_step
+#print 'one behind Harvard', one_behind_harvard[0][0]
+
+filter_distance -= 1
+child = unicode(one_behind_harvard[0][1])
+
+potential_parents = rdd.filter(lambda x: x[1] == filter_distance).partitionBy(256) 
+
+potential_parents_with_children = potential_parents.join(full_rdd2).mapValues(lambda x: x[1]) 
+
+parents_of_child = potential_parents_with_children.filter(lambda x: child in x[1]).keys() 
+
+second_step = parents_of_child.collect()
+
+two_behind_harvard = page_names.filter(lambda (K, V): V == int(second_step[0])).collect()

```

Comment:
Look at my comment above about a potentially cleaner way to approach this! At the very least, there is a loop-like structure here that would have made your code cleaner.